kerchunk/df.py

```diff
@@ -25,7 +28,182 @@ def _proc_raw(r):
+    """Write references into a parquet file.
+
+    Parameters
+    ----------
+    fname : str
+        Output filename.
+    json_obj : str, bytes, dict, list
+        JSON data for parquet file to be written.
+    """
+    json_obj = _normalize_json(json_obj)
+    with open(fname, "wb") as f:
+        f.write(json_obj)
+
+
 def refs_to_dataframe(
+    refs,
+    url,
+    storage_options=None,
+    row_group_size=1000,
+    compression="zstd",
+    engine="fastparquet",
+    **kwargs,
+):
+    """Write references as a store of parquet files with multiple row groups.
+    The directory structure should mimic a normal zarr store but instead of standard chunk
+    keys, references are saved as parquet dataframes with multiple row groups.
+
+    Parameters
+    ----------
+    refs: str | dict
+        Location of a JSON file containing references or a reference set already loaded
+        into memory. It will get processed by the standard referenceFS, to normalise
+        any templates, etc., it might contain.
+    url: str
+        Location for the output, together with protocol. This must be a writable
+        directory.
+    storage_options: dict | None
+        Passed to fsspec when for writing the parquet.
+    row_group_size : int
+        Number of references to store in each reference file (default 1000)
+    compression : str
+        Compression information to pass to parquet engine, default is zstd.
+    engine : {'fastparquet', 'pyarrow'}
+        Library to use for writing parquet files.
+    **kwargs : dict
+        Additional keyword arguments passed to parquet engine of choice.
+    """
+    if not os.path.exists(url):
+        os.makedirs(url)
+    if "refs" in refs:
+        refs = refs["refs"]
+    _write_json(
+        os.path.join(url, ".row_group_size"), dict(row_group_size=row_group_size)
+    )
+    fields = get_variables(refs)
+    for field in fields:
+        field_path = os.path.join(url, field)
+        if field.startswith("."):
+            # zarr metadata keys (.zgroup, .zmetadata, etc)
+            _write_json(field_path, refs[field])
+        else:
+            if not os.path.exists(field_path):
+                os.makedirs(field_path)
+            # Read the variable zarray metadata to determine number of chunks
+            zarray = ujson.loads(refs[f"{field}/.zarray"])
+            chunk_sizes = np.ceil(
+                np.array(zarray["shape"]) / np.array(zarray["chunks"])
+            )
+            if chunk_sizes.size == 0:
+                chunk_sizes = np.array([0])
+            nchunks = int(np.product(chunk_sizes))
+            extra_rows = row_group_size - nchunks % row_group_size
+            output_size = nchunks + extra_rows
+            # Initialize dataframe columns. Don't know why but int64 gives better
+            # compression ratio than int32
+            paths = np.full(output_size, np.nan, dtype="O")
+            offsets = np.zeros(output_size, dtype="int64")
```

Comment:
I noticed this in the comment but was wondering why I was getting better compression ratios setting this to `int64`, because it would definitely be ideal to use `int32` instead here.

Reply:
There could easily be chunks bigger than 4GB, let's leave it. For the compression ratio - I don't know, the algorithms are magic.